[HACK WEEK] Integrate periphery

RafaelKayumov opened the pull request:
## Description

Integrate [Periphery](https://github.com/peripheryapp/periphery) to run on Buildkite CI. Follows the approach introduced here: pbzQyC-3OH-p2

- Implements the `.periphery.yml` setup.
- Generates and uses `Scripts/Periphery/periphery_baseline.json` as filter list of existing legacy unused code issues.
- Implements the `Scripts/Periphery/setup-and-run-periphery.sh` to prepare and use the Periphery installation
- Implements the `.buildkite/commands/run-periphery.sh` as a Buildkite pipeline command. 
- Runs the `setup-and-run-periphery.sh` scan with `--skip-build` to reuse the build artifact from the build step
- Extends the `.buildkite/commands/build-for-testing.sh` step to also upload the `DerivedData/Index.noindex/DataStore/` data store index for the Periphery.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

RafaelKayumov committed:
Setup periphery.yml

RafaelKayumov committed:
Adjust periphery.yml for local runs

RafaelKayumov committed:
Add baseline feature to ignore the snapshot of existing issues

RafaelKayumov committed:
git ignore Periphery install

RafaelKayumov committed:
Setup periphery run script

RafaelKayumov committed:
Move baseline json to Scripts/Periphery folder from root

RafaelKayumov committed:
Add Periphery step in buildkite pipeline

RafaelKayumov committed:
Specify debug build config in build_for_testing lane

RafaelKayumov committed:
Add skip job type check for Periphery

RafaelKayumov committed:
Add gh commit status for periphery

RafaelKayumov committed:
Update artifact download

RafaelKayumov committed:
include Xcode Index Store in the build-products.tar artifact

RafaelKayumov committed:
Exclude test targets

RafaelKayumov committed:
Remove unnecessary chmod

RafaelKayumov committed:
Re-generate baseline

itsmeichigo reviewed (commented):
I'm a novice when it comes to app infrastructure tasks so I'll refrain from reviewing this PR. I'm just curious where I can get the reports from Periphery after this setup? (pardon the dumb question)

RafaelKayumov commented:
> I'm a novice when it comes to app infrastructure tasks so I'll refrain from reviewing this PR. I'm just curious where I can get the reports from Periphery after this setup? (pardon the dumb question)

You can go to the PR's Buildkite pipeline and check out the "Detect unused code" step. For example this is the one for this PR: https://buildkite.com/automattic/woocommerce-ios/builds/31502#019879ee-7dcd-453e-b5c2-1ed88921cf0f

If there are any unused code entries, the build will fail and the output will contain smth like:
```
woocommerce-ios/WooCommerce/Classes/Spotlight/SpotlightManager.swift:5:16: warning: Property 'unusedLetExample' is unused
```

AliSoftware commented on .buildkite/commands/run-periphery.sh:
```diff
@@ -0,0 +1,36 @@
+
+# Handle the result of the periphery scan
+if [[ "$TESTS_EXIT_STATUS" -ne 0 ]]; then
+  echo '😱 Unused code detected!'
```
💡  Idea for a future improvement for a future PR: I wonder if Periphery is able to write its results / logs into a file and we could use that to show the log output / list of unused code warnings in a Buildkite annotation, for easier access in the Buildkite UI than having to dig into the logs of the job 🤔 

This could be done by using something like `cat periphery-output.txt | buildkite-agent annotate --context periphery --style error` iinm, assuming `periphery-output.txt` being the file where periphery would write all the violations…

Though maybe it's not that complicated to look at the logs of the job after all 🤷 

Anyway, that's just food for thoughts for UI improvements later 🙃

RafaelKayumov replied:
Sounds interesting. I wonder if we could even route the output as a GH comment. Noted.

> Anyway, that's just food for thoughts for UI improvements later

Yeah, my initial idea was just to make it work and automate blocking the unused code in our PRs.

AliSoftware replied:
Technically if you're interested in looking at it as a future iteration later, I think that shouldn't be too complicated, as it seems `periphery scan` has a `--write-results <PATH>` option ideal for that (while still printing the errors to stdout as well, apparently 👍 .

So I think adopting this would just consist of:

- Adding the `--write-results periphery-errors.txt` flag to your call to `setup-and-run-periphery.sh` in line 18
- Add a call to `cat periphery-errors.txt | buildkite-agent annotate --context periphery --style error` (†) after this line 24 of that same script to put that content in a Buildkite annotation

(†) Or if you want to make it a tad fancier to wrap those into a small markdown, ``echo "## Periphery found unused code\n\n\`\`\`\n$(cat periphery-errors.txt)\n\`\`\`" | buildkite-agent annotate --context periphery --style error``

AliSoftware replied:
Actually created [a separate PR on top of this one](https://github.com/woocommerce/woocommerce-ios/pull/15984) to implement this myself, so we can test that separately if we wanted to and iterate 🙂

AliSoftware commented on Scripts/Periphery/setup-and-run-periphery.sh:
```diff
@@ -0,0 +1,54 @@
+if [ -x "$PERIPHERY_PATH" ]; then
+    echo "Executable found. Checking version..."
+    # Get the current installed version
+    CURRENT_VERSION=$($PERIPHERY_PATH version)
```
💄 nitpick: In case `$PERIPHERY_PATH` contains spaces (e.g. for someone who might have cloned the repo on their machine on a location that contains spaces and they want to run that script locally), you might want to add quotes around this

```suggestion
    CURRENT_VERSION=$("$PERIPHERY_PATH" version)
```

RafaelKayumov replied:
Done in [4b08069](https://github.com/woocommerce/woocommerce-ios/pull/15962/commits/4b08069ca0c65db258acc95ce4b52bcb72d0827f)

AliSoftware commented on Scripts/Periphery/setup-and-run-periphery.sh:
```diff
@@ -0,0 +1,54 @@
+    CURRENT_VERSION=$($PERIPHERY_PATH version)
+    
+    # Compare the current version with the desired version
+    if version_gt $PERIPHERY_VERSION $CURRENT_VERSION; then
```
💄 nitpick: Not sure that'd ever occur, but in case one day the `periphery version` command starts returning version strings with spaces (e.g. "3.2.1 beta 1"), better safe than sorry:
```suggestion
    if version_gt "$PERIPHERY_VERSION" "$CURRENT_VERSION"; then
```

RafaelKayumov replied:
done in [4b08069](https://github.com/woocommerce/woocommerce-ios/pull/15962/commits/4b08069ca0c65db258acc95ce4b52bcb72d0827f)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -815,6 +815,7 @@ platform :ios do
       workspace: WORKSPACE_PATH,
       scheme: TEST_SCHEME,
       derived_data_path: 'DerivedData',
+      xcargs: "-xcconfig 'config/WooCommerce.debug.xcconfig'",
```
❓  I'm not entirely familiar with how the `-xcconfig` flag works, i.e. does it completely overwrite any of the `.xcconfig` files that was set up at the Xcode project level, replacing all of them by this single `.xcconfig` intead? Or does it kinda "append" the content of this `xcconfig` to the ones defined by the Xcode project?

In other words, if the Xcode workspace + scheme + config defines its Build Settings to be set by `config/project.xcconfig` at project level, and `config/WooCommerce.xcconfig` at target level, and you add `-xcconfig 'config/WooCommerce.debug.xcconfig'` via the command lines here… will the settings from `config/project.xcconfig` set at workspace or project level in Xcode be ignored, or still included?

So I'm wondering if what you would instead need to fix your case wouldn't be to use `configuration: 'Debug'` rather than `xcargs: "-xcconfig 'config/WooCommerce.debug.xcconfig'"`? 🤔

AliSoftware replied:
I guess technically `WooCommerce.debug.xcconfig` already `#include "WooCommerce.common.xcconfig"` which itself `#include "Common.xcconfig"` so in practice we should be covered and ensure that all settings, even inherited ones, are still present. So that's why it works 👍 

But still, if one day we restructure our `.xcconfig` files in some different way (e.g. introduce a `Project.xcconfig` that we'd set at project level, then rely on Xcode have target-level build settings be automatically inherited from project-level build settings if not overriden at that target level), I wonder if that might break things (by having that `-xcconfig …` parameter also replace the xcconfig set at project level, thus making any settings from `Project.xcconfig` be ignored)… while `configuration: 'Debug'` would not (and arguably that's closer to what you actually want to do, i.e. enforce the use of debug config?)

RafaelKayumov replied:
Thx for looking into it.
It seems that specifying the `-xcconfig` overrides the project and target level configs. But I recall applying that override to make sure the "same conditions" are used for local and CI builds - I was struggling to achieve the same Periphery output. At the moment this override doesn't look as required to me. I'm going to remove it.

RafaelKayumov replied:
Removed in [7916cc4](https://github.com/woocommerce/woocommerce-ios/pull/15962/commits/7916cc436d413ee4fc79978b3f9c1e016c2031a4)

RafaelKayumov committed:
Use quotes for path and version

RafaelKayumov committed:
Delete -xcconfig override from run_tests call

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/252">23.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware commented on Scripts/Periphery/setup-and-run-periphery.sh:
```diff
@@ -0,0 +1,54 @@
+    
+    echo "Download and setup complete. Running periphery scan..."
+    # Run periphery scan with additional arguments
+    $PERIPHERY_PATH scan "$@"
```
💡  Should we add `--disable-update-check` to the flags (here + on line 49) given the script already takes care of the version check?

I also think that passing the `--relative-results` flag would be nice, to have relative paths in the output logs (instead of absolute paths that are referencing `/opt/ci/builds/builder/automattic/woocommerce-ios/` prefix).
Though that one I'm not sure if it makes more sense to hardcode it in this `.sh` script, or to pass it explicitly at call site from `run-periphery.sh`… but I think it'd make sense to have it in this `setup-and-run-periphery.sh` still because it makes sense for it to be a default we'd always want applied, both on CI and local runs?

```suggestion
    $PERIPHERY_PATH scan --disable-update-check "$@"
```
```suggestion
    $PERIPHERY_PATH scan --disable-update-check --relative-results "$@"
```

AliSoftware replied:
Implemented in [this PR stacked on top of this one](https://github.com/woocommerce/woocommerce-ios/pull/15984), so we can iterate independently if you wanted to unblock this one

AliSoftware commented on Scripts/Periphery/setup-and-run-periphery.sh:
```diff
@@ -0,0 +1,54 @@
+        update_and_run "$@"
+    else
+        echo "Current version ($CURRENT_VERSION) is up-to-date. Running periphery scan..."
+        $PERIPHERY_PATH scan "$@"
```
Same thought as above:

```suggestion
        $PERIPHERY_PATH scan --disable-update-check --relative-results "$@"
```

AliSoftware reviewed (approved):
Approving to unblock, just left a suggestions of adding 2 flags to the `periphery scan` to improve the output

[EDIT] Actually implemented those suggestions myself in a [separate PR](https://github.com/woocommerce/woocommerce-ios/pull/15984) so that you have all the options (merge this base PR independently and iterate on the other one later, or merge the other one in this one first if you like it before merging all to `trunk`… up to you)

AliSoftware committed:
Add Periphery errors as Buildkite annotation

AliSoftware committed:
Refactor setup-and-run-periphery.sh

Add --disable-update-check and --relative-results flags

DRY some code to avoid repetition of the commands to run scan

AliSoftware committed:
🚧 TO REVERT: Introduce violations on purpose

RafaelKayumov commented:
@AliSoftware Thx for contributions. I'll wait for your [PR](https://github.com/woocommerce/woocommerce-ios/pull/15984) before merging.

AliSoftware committed:
Revert "🚧 TO REVERT: Introduce violations on purpose"

This reverts commit 0203b2fc83ded0ffede7dedff724481af7a248ba.

AliSoftware committed:
Integrate periphery: add fancier reporting (#15984)

AliSoftware commented:
> I'm just curious where I can get the reports from Periphery after this setup? 

@itsmeichigo Quick update: following some improvements we made in https://github.com/woocommerce/woocommerce-ios/pull/15984, the CI job now _also_ posts a Buildkite annotation to report any Periphery warnings ([example here](https://buildkite.com/automattic/woocommerce-ios/builds/31537/annotations)) 🎉 
So that's now another (and arguably nicer) way to see those 🙂 _(though they're still in the job logs too if you prefer to look there)_
